src/core/dbus-service.c

```
@@ -532,6 +562,43 @@ static int bus_service_set_transient_property(
         if (streq(name, "StandardErrorFileDescriptor"))
                 return bus_set_transient_std_fd(u, name, &s->stderr_fd, &s->exec_context.stdio_as_fds, message, flags, error);
 
+        if (streq(name, "OpenFile")) {
+                char *path = NULL;
+                char *fdname = NULL;
+                uint64_t options;
+
+                r = sd_bus_message_enter_container(message, 'a', "(sst)");
+                if (r < 0)
+                        return r;
+
+                while ((r = sd_bus_message_read(message, "(sst)", &path, &fdname, &options)) > 0) {
+                        if (UNIT_WRITE_FLAGS_NOOP(flags)) continue;
+                        OpenFile *of = new(OpenFile, 1);
+                        if (!of)
+                                return -ENOMEM;
+
+                        *of = (OpenFile) {
+                                .path = strdup(path),
+                                .fdname = strdup(fdname),
+                                .flags = options,
+                        };
+                        if (!of->path || !of->fdname)
+                                return -ENOMEM;
+
+                        if (!path_is_absolute(of->path))
+                                return -EINVAL;
+
+                        LIST_APPEND(open_files, s->open_files, of);
+                        unit_write_settingf(u, flags|UNIT_ESCAPE_SPECIFIERS, name, "OpenFile=%s", open_file_to_string(of));
+                }
+
```

We should return any error that might happened in the `while` loop here

---

This wasn't actually resolved. The `while` loop can exit for two reasons:

1. There are no more entries in the array to read
2. An error occurred while reading one of the entries

If the second scenario happens, `r` will be negative and we need to make sure we return that error to the user.